File: src-javascript/engine/image/texture.js
"use strict";

var TEXTURE_IDS = 0;
const TEXTURE_POOL = new Map();
var texture_defer_enabled = false;


async function texture_init(src) {
    return texture_init_deferred(src, texture_defer_enabled);
}

async function texture_init_deferred(src, deffered) {
    let loaded_texture = await texture_get_from_global_pool(src);
    if (loaded_texture) {
        // This texture is already in use, bump the reference count and reuse it
        loaded_texture.references++;
        return loaded_texture;
    }

    let tex;
    try {
        tex = await fs_readimagebitmap(src);
    } catch (e) {
        console.error(e);
        return null;
    }

    let texture = texture_init_from_raw(tex.data, tex.size, false, tex.data.width, tex.data.height, 0, 0);

    // remember the filename to avoid duplicated textures
    texture.src_filename = await fs_get_full_path_and_override(src);

    // original size if was readed from an standard image file (png, jpg, bmp)
    texture.original_width = tex.original_width;
    texture.original_height = tex.original_height;

    if (!deffered) texture_upload_to_pvr(texture);

    return texture;
}

function texture_init_from_raw(ptr, size, in_vram, width, height, orig_width, orig_height) {
    let texture = {
        size: size,
        data_vram: null,
        data_ram: null,

        width: width,
        height: height,
        original_width: orig_width,
        original_height: orig_height,

        id: TEXTURE_IDS++,
        src_filename: null,
        references: 1,
        has_mipmaps: false
    };

    if (in_vram)
        texture.data_vram = ptr;
    else
        texture.data_ram = ptr;

    if (!in_vram && texture.data_ram instanceof DDS) {
        texture.has_mipmaps = ptr.Mipmaps.length > 0;
    }

    TEXTURE_POOL.set(texture.id, texture);

    return texture;
}


function texture_share_reference(texture) {
    texture.references++;
    return texture;
}

function texture_upload_to_pvr(texture) {
    if (texture.data_vram) return;

    if (!texture.data_ram) {
        // null data loaded
        return;
    }

    // C only
    /*
    texture->data_vram = pvr_mem_malloc(texture->size);
    if (!texture->data_vram) {
        logger_error("texture_upload_to_pvr() no enough video memory, required=%li available=%lu\n", texture->size, pvr_mem_available());
        return;
    }

    pvr_txr_load(texture->data_ram, texture->data_vram, texture->size);

    free(texture->data_ram);
    texture->data_ram = null;
    return;
    */

    texture.data_vram = webopengl_create_texture(pvr_context.webopengl, texture.width, texture.height, texture.data_ram);

    texture.data_ram.close();
    texture.data_ram = null;
}

function texture_get_size(texture) {
    return texture.size;
}

function texture_get_original_dimmensions(texture, output_dimmensions) {
    output_dimmensions[0] = texture.original_width;
    output_dimmensions[1] = texture.original_height;
    return output_dimmensions;
}

function texture_get_dimmensions(texture, output_dimmensions) {
    output_dimmensions[0] = texture.width;
    output_dimmensions[1] = texture.height;
    return output_dimmensions;

}


function texture_destroy_force(texture) {
    if (!texture) return;
    texture.references = 0;
    texture_destroy(texture);
}

function texture_destroy(texture) {
    if (!texture) return true;

    texture.references--;

    // check if the texture is shared somewhere
    if (texture.references > 0) {
        return false;
    }

    // C only
    /*
    
    if (texture->data_ram) {
        free(texture->data_ram);
    }
    if (texture->data_vram) {
        pvr_mem_free(texture->data_vram);
    }

    TEXTURE_POOL.delete(texture.id);

    string_destroy(texture->src_filename);

    free(texture);
    return true;

    */

    // JS only
    if (texture.ram) {
        texture.data_ram.close();
    }
    if (texture.data_vram) {
        webopengl_destroy_texture(pvr_context.webopengl, texture);
    }

    texture.ram = null;
    texture.vram = null;
    texture.src_filename = undefined;
    TEXTURE_POOL.delete(texture.id);
    return true;
}


async function texture_get_from_global_pool(src) {
    let texture = null;
    if (src) {
        let filename = await fs_get_full_path_and_override(src);
        for (let loaded_texture of TEXTURE_POOL.values()) {
            if (string_equals_ignore_case(loaded_texture.src_filename, filename)) {
                texture = loaded_texture;
                break;
            }
        }
        filename = undefined;
    }
    return texture;
}

function texture_disable_defering(disable) {
    texture_defer_enabled = !!disable;
}

File: src-desktop/kdy-e/engine/image/texture.cs
using System;
using System.Threading;
using Engine.Externals;
using Engine.Platform;
using Engine.Utils;

namespace Engine.Image;

public class Texture {
    private static int TEXTURE_IDS = 0;
    private static Map<Texture> TEXTURE_POOL = new Map<Texture>();
    private static bool texture_defer_enabled = true;

    private int size;
    internal WebGLTexture data_vram;
    private ImageData data_ram;

    internal int width;
    internal int height;
    internal int original_width;
    internal int original_height;

    private int id;
    internal string src_filename;
    private int references;
    private int cache_references;
    internal bool has_mipmaps;
    internal WebGLTexture[] data_vram_crhoma_planes;


    public static Texture Init(string src) {
        return Texture.InitDeferred(src, texture_defer_enabled);
    }

    public static Texture InitDeferred(string src, bool deffered) {
        Texture loaded_texture = Texture.GetFromGlobalPool(src);
        if (loaded_texture != null) {
            // This texture is already in use, bump the reference count and reuse it
            loaded_texture.references++;
            return loaded_texture;
        }

        ImageData data_ram = FS.ReadTexture(src);
        if (data_ram == null) return null;

        Texture texture = InitFromRAW(
            data_ram, data_ram.size, false, data_ram.pow2_width, data_ram.pow2_height, -1, -1
        );

        // remember the filename to avoid duplicated textures
        texture.src_filename = FS.GetFullPathAndOverride(src);

        // original size if was readed from an standard image file (png, jpg, bmp)
        texture.original_width = data_ram.orig_width;
        texture.original_height = data_ram.orig_height;

        if (Thread.CurrentThread.Name == "MainThread" && !deffered) texture.UploadToPVR();

        return texture;
    }

    public static Texture InitFromRAW(object ptr, int size, bool in_vram, int width, int height, int orig_width, int orig_height) {
        Texture texture = new Texture();
        texture.size = size;
        texture.data_vram = WebGLTexture.Null;
        texture.data_ram = null;

        texture.width = width;
        texture.height = height;
        texture.original_width = orig_width;
        texture.original_height = orig_height;

        texture.id = TEXTURE_IDS++;
        texture.src_filename = null;
        texture.references = 1;
        texture.cache_references = 0;
        texture.has_mipmaps = false;
        texture.data_vram_crhoma_planes = null;

        if (in_vram)
            texture.data_vram = (WebGLTexture)ptr;
        else
            texture.data_ram = (ImageData)ptr;

        if (!in_vram && texture.data_ram.pixelDataBuffer is DDSPixelDataBuffer) {
            DDSPixelDataBuffer dds = (DDSPixelDataBuffer)texture.data_ram.pixelDataBuffer;
            texture.has_mipmaps = dds.Mipmaps.Length > 0;
        }

        TEXTURE_POOL.Set(texture.id, texture);

        return texture;
    }


    public Texture ShareReference() {
        this.references++;
        return this;
    }

    public void Cache(bool adquire) {
        if (adquire)
            this.cache_references++;
        else if (this.cache_references > 0)
            this.cache_references--;

        //
        // This functionality is only available in the Dreamcast
        // OpenGL and WebGL has lot of limitations
        //
    }

    public void UploadToPVR() {
        if (!this.data_vram.IsNull) return;

        if (this.data_ram == null) {
            // null data loaded
            return;
        }

        this.data_vram = PVRContext.global_context.webopengl.CreateTexture(this.width, this.height, this.data_ram);

        if (this.data_vram.IsNull) {
            Logger.Error($"texture_upload_to_pvr() failed, id={this.id} src={this.src_filename}");
        }

        if (this.cache_references > 0) return;

        this.data_ram.Dispose();
        this.data_ram = null;
    }

    public int GetSize() {
        return this.size;
    }

    public void GetOriginalDimmensions(out float width, out float height) {
        width = this.original_width;
        height = this.original_height;
    }

    public void GetDimmensions(out float width, out float height) {
        width = this.width;
        height = this.height;

    }


    public void DestroyForce() {
        this.references = 0;
        this.cache_references = 0;
        Destroy();
    }

    public bool Destroy() {
        this.references--;

        // check if the texture is shared somewhere
        if (this.references > 0) {
            return false;
        }

        // JS only
        if (this.data_ram != null) {
            this.data_ram.Dispose();
        }
        if (!this.data_vram.IsNull) {
            PVRContext.global_context.webopengl.DestroyTexture(this.data_vram);
        }
        if (this.data_vram_crhoma_planes != null) {
            PVRContext.global_context.webopengl.DestroyTexture(this.data_vram_crhoma_planes[0]);
            PVRContext.global_context.webopengl.DestroyTexture(this.data_vram_crhoma_planes[1]);
        }

        this.data_ram = null;
        this.data_vram = WebGLTexture.Null;
        //free(this.src_filename);
        TEXTURE_POOL.Delete(this.id);
        return true;
    }


    public static Texture GetFromGlobalPool(string src) {
        Texture texture = null;
        if (src != null) {
            string filename = FS.GetFullPathAndOverride(src);
            foreach (Texture loaded_texture in TEXTURE_POOL) {
                if (loaded_texture.src_filename == filename) {
                    texture = loaded_texture;
                    break;
                }
            }
            //free(filename);
        }
        return texture;
    }

    public static void DisableDefering(bool disable) {
        Texture.texture_defer_enabled = disable;
    }

}
